galley/pkg/runtime/monitoring/monitoring.go

```diff
@@ -124,6 +143,39 @@ func RecordStateTypeCountWithContext(ctx context.Context, count int) {
 	}
 }
 
+// RecordDetailedStateType
+func RecordDetailedStateType(namespace, name string, collection fmt.Stringer, count int) {
+	collectionStr := strings.Split(collection.String(), "/")
+	// collection is of the format istio/<kind>/<version>/<name>
+	if len(collectionStr) < 4 {
+		log.Scope.Errorf("length of collection is less than 4, does not match expectation.")
+		return
+	}
+	ctx, err := tag.New(context.Background(), tag.Insert(NamespaceTag, namespace),
+		tag.Insert(NameTag, name), tag.Insert(VersionTag, collectionStr[2]))
+	if err != nil {
+		log.Scope.Errorf("Error creating monitoring context for counting state: %v", err)
+	} else {
+		RecordDetailedStateTypeWithContext(ctx, strings.Replace(collection.String(),
+			fmt.Sprintf("%s/", collectionStr[2]), "", 1), count)
+	}
+}
+
+// RecordDetailedStateTypeWithContext
+func RecordDetailedStateTypeWithContext(ctx context.Context, collection string, count int) {
+	if ctx == nil {
+		return
+	}
+	if stateTypeConfigTotal[collection] == nil {
```

Comment:
There is a race condition here. This method can potentially be called from multiple go-routines. Use sync.Map or  a RWMutex?

Reply:
Actually, sync.Map would not work, since you're also registering a view.

Reply:
Registering a view looks like is thread safe: tryRegisterView is using locks..
so used sync.Map only here..